src/tests/ef_vi/eflatency.c

```diff
@@ -99,34 +101,40 @@ const uint32_t raddr_he = 0xac010203;  /* 172.1.2.3 */
 const uint16_t port_he = 8080;
 
 
+static void checksum_udp_pkt(void * pkt_buf)
+{
+  ci_ether_hdr * const eth = (ci_ether_hdr*) pkt_buf;
+  ci_ip4_hdr * const ip4 = (void*) ((uintptr_t) eth + 14);
```

Comment:
`uintptr_t` implies to me that the (integer) result will be offset by a number of bytes, although `uint8_t*` or `char*` would work too.  Also note that the magic number `14` (copied from the original code) gets replaced with `ETH_HEADER_SIZE` in the commit that adds VLAN support.

Reply:
I will preface this comment by saying that I understand how incredibly pedantic this note is and indeed, as you have mentioned, any of these will really work in practice. Also, apologies if the below is all stuff that you know - I only learnt the difference between some of these things a few months ago and thought the difference was quite fascinating!

I think my hesitance is mainly because the C standard doesn't really define arithmetic on `uintptr_t`, as far as my understanding goes. The standard itself only defines the type such that a valid `void*`, cast to `uintptr_t`, then cast back into a `void*` will necessarily be the same as the original pointer.

In the case of pointer arithmetic, I think it's best to define the exact type/unit you're traversing in memory. At the time of the second commit (the diff that the original comment is on), the magic `14` is used as the length of an ethernet frame (without VLAN tagging). In that situation, we want to traverse exactly 14 octets, so we should use a `uint8_t*`.

However, because this is later changed to drop the magic `14` and replace it with a conditional `sizeof` operation, I would argue that `uint8_t*` is no longer the appropriate unit. The unit of the data returned by `sizeof` is defined to be a byte, which means we should be using a `char*`.

All that being said, I notice you have changed this to use `uint8_t*`, and I imagine we would be in a far stickier situation than eflatency having dodgy headers if compiled on a machine where 1 byte isn't 8 bits.